performance-tracking/src/main/java/com/rakuten/tech/mobile/perf/runtime/internal/ConfigStore.java

```diff
@@ -93,72 +109,94 @@ public void run() {
+            }
+          });
+
     } catch (PackageManager.NameNotFoundException e) {
       Log.d(TAG, "Error building request to config API", e);
     }
+  }
 
-    if (param != null) {
-      new ConfigurationRequest(
-              urlPrefix,
-              subscriptionKey,
-              param,
-              new Response.Listener<ConfigurationResult>() {
-                @Override
-                public void onResponse(ConfigurationResult newConfig) {
-                  if (newConfig == null && Tracker.isTrackerRunning()) {
-                    TrackingManager.deinitialize();
-                  }
-
-                  ConfigurationResult prevConfig = readConfigFromCache();
-                  boolean shouldRollDice =
-                      (newConfig != null && Tracker.isTrackerRunning() && prevConfig == null)
-                          || (prevConfig != null
-                              && newConfig != null
-                              && newConfig.getEnablePercent() < prevConfig.getEnablePercent());
-
-                  if (shouldRollDice) {
-                    double randomNumber =
-                        new Random(System.currentTimeMillis()).nextDouble() * 100.0;
-                    if (randomNumber > newConfig.getEnablePercent()) {
-                      // DeInitialize Tracking Manager
-                      TrackingManager.deinitialize();
-                    }
-                  }
-                  writeConfigToCache(newConfig);
-                  getObservable().publish(newConfig);
-                }
-              },
-              new Response.ErrorListener() {
-                @Override
-                public void onErrorResponse(VolleyError error) {
-                  // DeInitialize Tracking Manager as we couldn't able to get new config from api
-                  TrackingManager.deinitialize();
-                  Log.d(TAG, "Error loading configuration", error);
-                }
-              })
-          .queue(requestQueue);
+  private void handleNewConfig(@Nullable ConfigurationResponse newConfig) {
+    if (newConfig == null && Tracker.isTrackerRunning()) {
+      TrackingManager.deinitialize();
     }
-  }
 
-  private void writeConfigToCache(ConfigurationResult result) {
-    prefs.edit().putString(CONFIG_KEY, new Gson().toJson(result)).apply();
+    ConfigurationResponse prevConfig = readConfigFromCache();
+    boolean shouldRollDice =
+        (newConfig != null && Tracker.isTrackerRunning() && prevConfig == null)
+            || (prevConfig != null
+            && newConfig != null
+            && newConfig.getEnablePercent() < prevConfig.getEnablePercent());
+
+    if (shouldRollDice) {
+      double randomNumber =
+          new Random(System.currentTimeMillis()).nextDouble() * 100.0;
+      if (randomNumber > newConfig.getEnablePercent()) {
+        // DeInitialize Tracking Manager
+        TrackingManager.deinitialize();
+      }
+    }
+    if(newConfig != null) {
```

Comment:
It looks like `shouldRollDice` is always false when `newConfig` is `null`, so you could move the parts related to the random number generation into this conditional and then simplify the logic for `shouldRollDice`.

- [ ] Refactor to minimize `newConfig` null check

Reply:
I actually like the current sequential structure better, now we have
```
private void handleNewConfig(@Nullable ConfigurationResponse newConfig) {
    // (1) stop operation
    if (newConfig == null && Tracker.isTrackerRunning()) {
      TrackingManager.deinitialize();
    }

    // (2) flip dice & (maybe) stop
    ConfigurationResponse prevConfig = readConfigFromCache();
    boolean shouldRollDice =
        (newConfig != null && Tracker.isTrackerRunning() && prevConfig == null)
            || (prevConfig != null
            && newConfig != null
            && newConfig.getEnablePercent() < prevConfig.getEnablePercent());

    if (shouldRollDice) {
      double randomNumber =
          new Random(System.currentTimeMillis()).nextDouble() * 100.0;
      if (randomNumber > newConfig.getEnablePercent()) {
        // DeInitialize Tracking Manager
        TrackingManager.deinitialize();
      }
    }
    
    // (3) update local cache
    if(newConfig != null) {
      prefs.edit().putString(CONFIG_KEY, newConfig.toString()).apply();
    } else {
      prefs.edit().remove(CONFIG_KEY).apply();
    }
    // (4) notify observers stuff
    getObservable().publish(newConfig);
  }
```

se we end up with this sequence of 

* stop operation 
* flip dice & (maybe) stop
* update local cache
* notify observers stuff

I think we shouldn't mangle these distinct cases into nested conditionals.

I agree that `shouldRollDice` is complicated, but by nesting it deeper in conditionals the complexity is only hidden by indirection, you still need to keep that context in mind, i.e. the mental load doesn't decrease.

what we could do instead is refactor the boolean expression by rearranging the boolean algebra, what do you think of this instead?

```
boolean shouldRollDice = newConfig != null &&
        ((Tracker.isTrackerRunning() && prevConfig == null) 
            || (prevConfig != null && newConfig.getEnablePercent() < prevConfig.getEnablePercent())
        );
```

Reply:
Yeah, that makes sense. It's better with the refactored boolean.